### PR TITLE
Support cross-ref from `xr` package

### LINKED ISSUE #3391: Completion of references from external document (`xr` package)
### Disable all the other extensions except for LaTeX Workshop, restart VS Code, and check that you can not see the requested feature. [Required]

You still see this issue?: Yes

### Make sure to visit the [wiki FAQ](https://github.com/James-Yu/LaTeX-Workshop/wiki/FAQ) before requesting a feature.

You visited the wiki?: Yes

## Is your requested feature  related to a problem? Please describe. [Required]

When a document is setup to use references from another compiled LaTeX document using `xr`, the references work at compile time but are not suggested by autocomplete.

```
\usepackage{xr}
\externaldocument{other_document}
```

## To Reproduce  [Required]

1. Create a valid document `a.tex` with a label `\label{sec:example}`. Compile it.
2. Create a valid document `b.tex`. Use package `xr`'s `\externaldocument{a}` command to include labels from `a.tex`. Add a reference to `\ref{sec:example}`. Compile.
3. The document compiles correctly but label `sec:example` was not included in the autocomplete for `\ref{` in document `b.tex`.

## Describe the solution you'd like  [Required]

The set of all labels (sections, figures, equations, etc) would be included in the autocompletion of `\ref{` and `\autoref{`.
Ideally, there would be a new option in the extention configuration to enable or disable this.

## Desktop [Required]

Please write exact version numbers. Please don't write `latest` instead of exact numbers.

 - OS: macOS 11.6
 - VS Code version: 1.69.2
 - Extension version: 8.27.2
 - TeX distribution version: MacTex 2021

### Are you using VSCodium?

No

### Are you using the Snap or Flatpack versions of VS Code?

No

### Are you using LaTeX Workshop with VS Code Remote?

Yes (but the same issue occurs locally).

### PR DESCRIPTION
**Pull request by James-Yu**
This PR resolves #3391 by following the `\externaldocument` command provided by `xr` package.

External documents are cached in a new entry `external` in the `context` of each TeX file, and they are also watched. Upon providing references, both `includedTeX` and `external` will be iterated to create a complete list.

The optional `prefix` of `\externaldocument` is also parsed and provided in the intellisense.

A test is added to test the feature.

### CODE REVIEW

**Review by jlelong — CHANGES_REQUESTED** [review]

File: src/components/cacher.ts [hunk 1]
```diff
@@ -15,21 +15,16 @@ import { PathUtils } from './cacherlib/pathutils'
...
 
 const logger = getLogger('Cacher')
 
 export interface Context {
-    /**
-     * The dirty (under editing) content of the LaTeX file if opened in vscode,
-     * the content on disk otherwise.
-     */
+    // The dirty (under editing) content of the LaTeX file if opened in vscode,
+    // the content on disk otherwise.
```
[the remark this answers is not in the document]
Comment (James-Yu) [reply src/components/cacher.ts:25]: I was too unfamiliar with the comment things in the past years. Will fix definitely!

File: src/components/cacher.ts [hunk 2]
```diff
@@ -158,8 +153,31 @@ export class Cacher {
...
+        while (true) {
+            const result = externalDocRegExp.exec(contentTrimmed)
+            if (!result) {
+                break
+            }
+
+            const texDirs = vscode.workspace.getConfiguration('latex-workshop').get('latex.texDirs') as string[]
+            const externalPath = utils.resolveFile([path.dirname(filePath), path.dirname(rootPath), ...texDirs], result[2])
+            if (!externalPath || !fs.existsSync(externalPath) || path.relative(externalPath, rootPath) === '') {
+                continue
```
[the remark this answers is not in the document]
Comment (James-Yu) [reply src/components/cacher.ts:169]: Will do!

File: src/components/cacher.ts [hunk 3]
```diff
@@ -158,8 +153,31 @@ export class Cacher {
...
+            }
+
+            this.contexts[rootPath].external[externalPath] = result[1] || ''
+            logger.log(`External document ${externalPath} from ${filePath} .`)
```
[the remark this answers is not in the document]
Comment (James-Yu) [reply src/components/cacher.ts:173]: Will do!

File: src/providers/completer/reference.ts [hunk 4]
```diff
@@ -97,13 +91,50 @@ export class Reference implements IProvider {
...
+            const startSize = included.size
+            included.forEach(cachedFile => {
+                lw.cacher.getIncludedTeX(cachedFile).forEach(includedTeX => {
+                    if (includedTeX === cachedFile) {
+                        return
+                    }
+                    included.add(includedTeX)
+                    if (prefixes[includedTeX]) {
+                        delete prefixes[includedTeX]
+                    }
```
[the remark this answers is not in the document]
Comment (James-Yu) [reply src/providers/completer/reference.ts:111]: I agree this is a be convoluted. I'm yet to figure a better way of implementation. Will comment carefully.

p.s. It's years of coding alone in the past, letting me forget about comment and readability. Apologize for that and will recover very fast!

File: src/providers/completer/reference.ts [hunk 5]
```diff
@@ -8,17 +8,11 @@ import {computeFilteringRange} from './completerutils'
...
     file: string,
-    /**
-     * The position that defines the ref.
-     */
+    // The position that defines the ref.
     position: vscode.Position,
-    /**
-     *  Stores the ref number.
-     */
+    //  Stores the ref number.
```
Comment (jlelong) [inline src/providers/completer/reference.ts:15]: Same as above

**Review by jlelong — APPROVED** [review]

**Pull request merged**